src/Disks/IO/CachedOnDiskWriteBufferFromFile.cpp

```diff
@@ -145,10 +144,9 @@ FileSegment & FileSegmentRangeWriter::allocateFileSegment(size_t offset, FileSeg
 
     /// We set max_file_segment_size to be downloaded,
     /// if we have less size to write, file segment will be resized in complete() method.
-    auto holder = cache->set(key, offset, cache->getMaxFileSegmentSize(), create_settings);
-    chassert(holder->size() == 1);
-    holder->moveTo(file_segments);
-    return file_segments.back();
+    file_segments = cache->set(key, offset, cache->getMaxFileSegmentSize(), create_settings);
```

Comment:
the point is in deleting this code with `moveTo` logic, previously all file segments which were written to cache in write-through cache during query were hold in FileSegmentsHolder till the end of the query, e.g. they were non-releasable, but the insert query can last for hours, thus it will hold a enormous amount of unreleasable file segments in cache.

Reply:
in other words we kept inserting new segments into the end of `file_segments` and now we remove its content and starting from scratch?

Reply:
interesting how much memory we could accumulate this way. maybe inserts became significantly less memory hungry?)

Reply:
> in other words we kept inserting new segments into the end of file_segments and now we remove its content and start from scratch?

yeah, e.g. `FileSegmentsHolder` now will always hold at most 1 file segment - the one that currently is being downloaded. For read-through cache it is already done - file segments are removed from holder once processed, do not know why I did not do the same for write-through cache.

> interesting how much memory we could accumulate this way. maybe inserts became significantly less memory hungry?)

possibly yes, though I doubt it was so significant, it was always 1 file segment per 100mb of data + there was an upper limit for the number of file segments created during an insert: as before this pr file segments were hold in a holder, they were non-releasable, so an insert could hold in memory in total at most all file segments which can fit in cache according to limits.